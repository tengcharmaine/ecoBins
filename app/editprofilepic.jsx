import { Text, Button, IconButton } from "react-native-paper"
import { StyleSheet, View, Image, TouchableOpacity, ActivityIndicator } from "react-native";
import { useNavigation, useFocusEffect } from "expo-router";
import * as ImagePicker from 'expo-image-picker';
import React, { useState } from "react";
import { supabase } from "../lib/supabase";
import { Ionicons } from '@expo/vector-icons';
import { useFonts } from 'expo-font';


export default function editprofilepic() {

    const [loaded] = useFonts({
        Poppins: require('../assets/fonts/Poppins-Regular.ttf'),
        PoppinsBold: require('../assets/fonts/Poppins-Bold.ttf'),
        PoppinsSemiBold: require('../assets/fonts/Poppins-SemiBold.ttf'),
        PoppinsBlack: require('../assets/fonts/Poppins-Black.ttf'),
    });

    const styles = StyleSheet.create({
        container: {
            flex: 1, 
            justifyContent: 'center',  
            alignItems: 'center',
            backgroundColor: "white",

        },
        input: {
            borderColor: "black",
            borderWidth: 1,
            backgroundColor: "white",
            width: '75%',
            borderRadius: 5
        },
        button: {
            borderColor: "black",
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: "#c7dede",
            width: '85%',
            height: 60,
            marginTop: 20,
            marginBottom: 10,
            borderRadius: 20,
        },
        defaultButton: {
            borderColor: "black",
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: "#dedede",
            width: '85%',
            height: 60,
            marginTop: 20,
            marginBottom: 10,
            borderRadius: 20, 
        },

        text: {
            color: "black",
            fontWeight: 'bold',
            fontSize: 19,
            textAlign: 'center',
            fontFamily: 'PoppinsSemiBold',
            justifyContent: 'center'
        },

        text1: {
            marginTop: 20,
            color: "black",
            marginBottom: 10,
            fontSize: 15,
            textAlign: 'center',
            width: '70%',
            fontFamily: 'PoppinsSemiBold',
        },

        title: {
            color: "black",
            marginTop: 20,
            justifyContent: 'center',
            fontWeight: 'bold',
            fontSize: 34,
            textAlign: 'left',
            marginRight: 40,
            marginBottom: 15,
            fontFamily: 'PoppinsBold',
        },

        button2: {
            borderColor: "black",
            borderStyle: 'dashed',
            borderWidth: 1,
            backgroundColor: "#dedede",
            width: '80%',
            height: '45%',
            marginTop: 60,
            marginLeft: 10,
            borderRadius: 10,
            alignItems: 'center',
            justifyContent: 'center',
            flexDirection: 'column',
        },

        error: {
            color: "red",
            fontFamily: 'Poppins',
            marginTop: 10,
        },

        backButton: {
            position: 'absolute',
            top: 60,
            left: 20,
            zIndex: 1,
        },
        uploadImage: {
            width: 90,
            height: 90,
            resizeMode: 'contain',
            alignSelf: 'center',
        },
    });

    const [image, setImage] = useState(null);
    const [error, setError] = useState(null);
    const [showButton, setShowButton] = useState(true);
    const [loading, setLoading] = useState(false);
    const navigation = useNavigation();

    useFocusEffect(
        React.useCallback(() => {
            // Reset the image state when the screen is focused
            setImage(null);
            setShowButton(true);
        }, [])
    );

    const resetToDefault = async () => {
        const { data: { user } } = await supabase.auth.getUser();
        const { data, error } = await supabase
                    .from('users')
                    .update({ profile: 'https://upload.wikimedia.org/wikipedia/commons/thumb/2/2c/Default_pfp.svg/1200px-Default_pfp.svg.png' })
                    .eq('id', user.id)
                    .single();
        navigation.goBack();

    };
    
    const handleAddImage = async () => {
            let result = await ImagePicker.launchImageLibraryAsync({
                mediaTypes: ImagePicker.MediaTypeOptions.Images,
                allowsEditing: true,
                quality: 1,
                base64: true,
                exif: true,
            });
            if (!result.canceled) {
                if (result.assets[0].uri) {
                    setImage(result.assets[0].uri);
                } else if (result.base64) {
                    setImage(`data:image/jpeg;base64,${result.base64}`);
                }
                setShowButton(false);
            }
    };
    

    const deleteImage = () => {
        setImage(null);
        setShowButton(true);
    };
  
    const saveImage = async () => {
      setLoading(true);
        if (image) {
          try {
            const { data: { user } } = await supabase.auth.getUser();
            console.log('User ID:', user.id);
      
            // Upload the image file to Supabase storage
            const { data: uploadedData, error: uploadError } = await supabase.storage
              .from('profilepic')
              .upload(`${new Date().getTime()}`, { uri: image, type: 'jpg', name: 'name.jpg' });
      
            if (uploadError) {
              setError('Error uploading profile picture: ' + uploadError.message);
              return;
            }
      
            console.log(uploadedData);

            // Get the public URL of the uploaded image
            const { data: publicUrl, error: getUrlError } = await supabase.storage
              .from('profilepic')
              .getPublicUrl(uploadedData.path);
            
            if (getUrlError) {
              setError('Error retrieving public URL: ' + getUrlError.message);
              return;
            }
      
            const uploadedImage = publicUrl;
            console.log(publicUrl);
      
            // Update the user's profile picture URL in the users table
            const { data: updateUserData, error: updateUserError } = await supabase
              .from('users')
              .update({ profile: uploadedImage.publicUrl })
              .eq('id', user.id)
              .single();
      
            if (updateUserError) {
              setError('Error updating user profile picture: ' + updateUserError.message);
              return;
            }
      
            console.log('Profile picture updated successfully!');
            setLoading(false);

            navigation.goBack();
            
          } catch (error) {
            setError('Error updating profile picture: ' + error.message);
          }
        } else {
          setError('No image selected. Please choose a profile picture.');
        }
      };
      
      const goBack = () => {
        navigation.navigate('settings');
      }
      
      if (!loaded) {
        return (
          <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
            <ActivityIndicator size="large" color="black" />
          </View>
        );
      }

    return (
        <View style={styles.container}>
            <TouchableOpacity onPress={goBack} style={styles.backButton}>
                <Ionicons name="arrow-back" size={24} color="black" />
            </TouchableOpacity>
            {/* <Text style={styles.title}>Change profile picture</Text> */}
            {image && (<View>
                <Image source={{ uri: image }} style={{ width: 270, height: 270 }} />
                <IconButton
                    icon="delete"
                    size={30}
                    color="black"
                    onPress={deleteImage}
                    style={{ position: 'absolute', top: 0, right: 0 }}
                />
            </View>)}
            {showButton && (
                <TouchableOpacity onPress={handleAddImage} style={styles.button2}>
                    <Image source={require("../images/profile.png")} style={styles.uploadImage} />
                    <Text style={styles.text1}>Choose a new profile picture</Text>
                </TouchableOpacity>
            )}
            {error !== null && <Text style={styles.error}>{error}</Text>}
            <TouchableOpacity style={styles.button} onPress={saveImage}>
                {loading ? <Text style={styles.text}>Loading...</Text> : <Text style={styles.text}>Confirm</Text>}
            </TouchableOpacity>
            <TouchableOpacity style={styles.defaultButton} onPress={resetToDefault}>
                <Text style={styles.text}>Reset to Default</Text>
            </TouchableOpacity>
        </View> 
    );
}
