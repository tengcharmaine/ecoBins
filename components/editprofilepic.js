import { Text, Button, IconButton } from "react-native-paper"
import { StyleSheet, View, Image, TouchableOpacity } from "react-native";
// import { useNavigation, useFocusEffect } from "expo-router";
import * as ImagePicker from 'expo-image-picker';
import React, { useState } from "react";
import { supabase } from "../lib/supabase";
// import { Ionicons } from '@expo/vector-icons';

export default function editprofilepic() {
    const styles = StyleSheet.create({
        container: {
            flex: 1, 
            justifyContent: 'center',  
            alignItems: 'center',
        },
        input: {
            borderColor: "black",
            borderWidth: 1,
            backgroundColor: "white",
            width: '75%',
            borderRadius: 5
        },
        button: {
            borderColor: "black",
            alignItems: 'center',
            backgroundColor: "#c7dede",
            width: '25%',
            marginTop: 20,
            marginBottom: 10,
            borderRadius: 10,
            
        },
        defaultButton: {
            borderColor: "black",
            alignItems: 'center',
            backgroundColor: "#c7dede",
            width: '40%',
            marginTop: 20,
            marginBottom: 10,
            borderRadius: 10,
            
        },

        text: {
            color: "black",
            marginTop: 10,
            marginBottom: 10,
            fontSize: 15,
            textAlign: 'center',
            width: '70%',
        },

        text1: {
            color: "black",
            marginTop: 20,
            textAlign: 'left',
            marginRight: 230,
            marginBottom: 5,
        },

        title: {
            color: "black",
            fontSize: 20,
            marginBottom: 20, 
            fontWeight: "bold",
        },

        button2: {
            borderColor: "black",
            backgroundColor: "#cccc",
            width: '75%',
            height: '40%',
            marginTop: 20,
            marginLeft: 10,
            borderRadius: 10,
            position: 'relative',
            alignItems: 'center',
            justifyContent: 'center',
            flexDirection: 'row',
        },

        error: {
            color: "red",
        },

        backButton: {
            position: 'absolute',
            top: 60,
            left: 20,
            zIndex: 1,
        },
    });

    const [image, setImage] = useState(null);
    const [error, setError] = useState(null);
    const [showButton, setShowButton] = useState(true);
    // const navigation = useNavigation();

    useFocusEffect(
        React.useCallback(() => {
            // Reset the image state when the screen is focused
            setImage(null);
            setShowButton(true);
        }, [])
    );

    const resetToDefault = async () => {
        const { data: { user } } = await supabase.auth.getUser();
        const { data, error } = await supabase
                    .from('users')
                    .update({ profile: 'https://upload.wikimedia.org/wikipedia/commons/thumb/2/2c/Default_pfp.svg/1200px-Default_pfp.svg.png' })
                    .eq('id', user.id)
                    .single();
        navigation.goBack();

    };
    
    const handleAddImage = async () => {
            let result = await ImagePicker.launchImageLibraryAsync({
                mediaTypes: ImagePicker.MediaTypeOptions.Images,
                allowsEditing: true,
                quality: 1,
                base64: true,
                exif: true,
            });
            if (!result.canceled) {
                if (result.assets[0].uri) {
                    setImage(result.assets[0].uri);
                } else if (result.base64) {
                    setImage(`data:image/jpeg;base64,${result.base64}`);
                }
                setShowButton(false);
            }
    };
    

    const deleteImage = () => {
        setImage(null);
        setShowButton(true);
    };
  
    const saveImage = async () => {
        if (image) {
          try {
            const { data: { user } } = await supabase.auth.getUser();
            console.log('User ID:', user.id);
      
            // Upload the image file to Supabase storage
            const { data: uploadedData, error: uploadError } = await supabase.storage
              .from('profilepic')
              .upload(`${new Date().getTime()}`, { uri: image, type: 'jpg', name: 'name.jpg' });
      
            if (uploadError) {
              setError('Error uploading profile picture: ' + uploadError.message);
              return;
            }
      
            console.log(uploadedData);

            // Get the public URL of the uploaded image
            const { data: publicUrl, error: getUrlError } = await supabase.storage
              .from('profilepic')
              .getPublicUrl(uploadedData.path);
            
            if (getUrlError) {
              setError('Error retrieving public URL: ' + getUrlError.message);
              return;
            }
      
            const uploadedImage = publicUrl;
            console.log(publicUrl);
      
            // Update the user's profile picture URL in the users table
            const { data: updateUserData, error: updateUserError } = await supabase
              .from('users')
              .update({ profile: uploadedImage.publicUrl })
              .eq('id', user.id)
              .single();
      
            if (updateUserError) {
              setError('Error updating user profile picture: ' + updateUserError.message);
              return;
            }
      
            console.log('Profile picture updated successfully!');
            navigation.goBack();
            
          } catch (error) {
            setError('Error updating profile picture: ' + error.message);
          }
        } else {
          setError('No image selected. Please choose a profile picture.');
        }
      };
      
    //   const goBack = () => {
    //     navigation.navigate('settings');
    //   }
      
    return (
        <View style={styles.container}>
            <TouchableOpacity onPress={goBack} style={styles.backButton}>
                {/* <Ionicons name="arrow-back" size={24} color="black" /> */}
            </TouchableOpacity>
            <Text style={styles.title}>Change profile picture</Text>
            {image && (<View>
                <Image source={{ uri: image }} style={{ width: 270, height: 270 }} />
                <IconButton
                    icon="delete"
                    size={30}
                    color="black"
                    onPress={deleteImage}
                    style={{ position: 'absolute', top: 0, right: 0 }}
                />
            </View>)}
            {showButton && (
                <Button onPress={handleAddImage} style={styles.button2}>
                    <Text style={styles.text}>Choose new profile picture</Text>
                </Button>
            )}
            {error !== null && <Text style={styles.error}>{error}</Text>}
            <Button style={styles.button} onPress={saveImage}>
                <Text style={styles.text}>Confirm</Text>
            </Button>
            <Button style={styles.defaultButton} onPress={resetToDefault}>
                <Text style={styles.text}>Reset to Default</Text>
            </Button>
        </View> 
    );
}