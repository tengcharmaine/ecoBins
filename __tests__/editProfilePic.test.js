import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import editprofilepic from '../components/editprofilepic';

jest.mock("expo-image-picker", () => ({
  launchImageLibraryAsync: jest.fn(() => Promise.resolve({ canceled: false, assets: [{ uri: "mocked-image-uri" }] })),
}));

jest.mock("../lib/supabase", () => ({
  supabase: {
    auth: {
      getUser: jest.fn(() => Promise.resolve({ data: { user: { id: "mocked-user-id" } } })),
    },
    storage: {
      from: () => ({
        upload: jest.fn(() => Promise.resolve({ data: { path: "mocked-uploaded-path" } })),
        getPublicUrl: jest.fn(() => Promise.resolve({ data: { publicUrl: "mocked-public-url" } })),
      }),
    },
    from: () => ({
      update: jest.fn(() => Promise.resolve({ data: { profile: "mocked-profile-url" } })),
    }),
  },
}));

// jest.mock("@react-navigation/compat", () => ({
//   withNavigation: (Component) => (props) => <Component {...props} navigation={{ navigate: jest.fn() }} />,
// }));

describe('EditProfilePic', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<editprofilepic />).toJSON();

    // Check if the required elements are rendered
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with an initial image', () => {
    // Set an initial value for the image state
    const initialImage = 'https://example.com/default-profile-pic.jpg';

    // Render the component with the initialImage as a prop
    const tree = renderer.create(<editprofilepic image={initialImage} />).toJSON();

    // Check if the required elements are rendered
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with an error message', () => {
    // Set an error message for the error state
    const errorMessage = 'Error uploading image';

    // Render the component with the errorMessage as a prop
    const tree = renderer.create(<editprofilepic error={errorMessage} />).toJSON();

    // Check if the required elements are rendered, including the error message
    expect(tree).toMatchSnapshot();
  });

});
