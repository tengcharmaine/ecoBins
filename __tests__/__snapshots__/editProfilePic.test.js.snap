// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfilePic renders correctly 1`] = `<editprofilepic />`;

exports[`EditProfilePic renders correctly with an error message 1`] = `
<editprofilepic
  error="Error uploading image"
/>
`;

exports[`EditProfilePic renders correctly with an initial image 1`] = `
<editprofilepic
  image="https://example.com/default-profile-pic.jpg"
/>
`;
